File: core/src/main/java/dev/vibris/core/PendingSourceRoot.java
package dev.vibris.core;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.attribute.BasicFileAttributes;

import static java.nio.file.LinkOption.NOFOLLOW_LINKS;

final class PendingSourceRoot {
    private final Path root;

    PendingSourceRoot(Path root) {
        this.root = root.toAbsolutePath().normalize();
    }

    void prepare() throws Failure {
        try {
            OwnedPathIdentity.createDirectoriesSafely(root);
            requireOrdinaryAncestors();
        } catch (IOException exception) {
            throw new Failure("Pending source root is missing or unsafe.", exception);
        }
        clear();
    }

    void clear() throws Failure {
        try {
            requireOrdinaryAncestors();
            try (var children = Files.list(root)) {
                for (Path child : children.toList()) {
                    if (!deleteEntry(child)) {
                        throw new Failure("Pending source root contains an unsafe entry.");
                    }
                }
            }
        } catch (IOException exception) {
            throw new Failure("Failed to clear the pending source root.", exception);
        }
    }

    private static boolean deleteEntry(Path child) throws IOException {
        BasicFileAttributes attributes = Files.readAttributes(child, BasicFileAttributes.class, NOFOLLOW_LINKS);
        if (attributes.isSymbolicLink() || attributes.isOther() || Files.isSymbolicLink(child)) {
            Files.delete(child);
            return true;
        }
        return OwnedSourceTree.delete(child);
    }

    private void requireOrdinaryAncestors() throws IOException {
        Path current = root;
        while (current != null) {
            OwnedPathIdentity.captureDirectory(current);
            current = current.getParent();
        }
    }

    static final class Failure extends Exception {
        Failure(String message) {
            super(message);
        }

        Failure(String message, Throwable cause) {
            super(message, cause);
        }
    }
}

File: core/src/main/kotlin/dev/vibris/core/PendingSourceRoot.kt
package dev.vibris.core

import java.io.IOException
import java.nio.file.Files
import java.nio.file.LinkOption.NOFOLLOW_LINKS
import java.nio.file.Path
import java.nio.file.attribute.BasicFileAttributes

internal class PendingSourceRoot(root: Path) {
    private val root = root.toAbsolutePath().normalize()

    @Throws(Failure::class)
    fun prepare() {
        try {
            OwnedPathIdentity.createDirectoriesSafely(root)
            requireOrdinaryAncestors()
        } catch (exception: IOException) {
            throw Failure("Pending source root is missing or unsafe.", exception)
        }
        clear()
    }

    @Throws(Failure::class)
    fun clear() {
        try {
            requireOrdinaryAncestors()
            Files.list(root).use { children ->
                for (child in children.toList()) {
                    if (!deleteEntry(child)) {
                        throw Failure("Pending source root contains an unsafe entry.")
                    }
                }
            }
        } catch (exception: IOException) {
            throw Failure("Failed to clear the pending source root.", exception)
        }
    }

    @Throws(IOException::class)
    private fun deleteEntry(child: Path): Boolean {
        val attributes = Files.readAttributes(child, BasicFileAttributes::class.java, NOFOLLOW_LINKS)
        if (attributes.isSymbolicLink || attributes.isOther || Files.isSymbolicLink(child)) {
            Files.delete(child)
            return true
        }
        return OwnedSourceTree.delete(child)
    }

    @Throws(IOException::class)
    private fun requireOrdinaryAncestors() {
        var current: Path? = root
        while (current != null) {
            OwnedPathIdentity.captureDirectory(current)
            current = current.parent
        }
    }

    class Failure @JvmOverloads constructor(
        message: String,
        cause: Throwable? = null,
    ) : Exception(message, cause)
}